engine/crates/graphql-schema-diff/src/state.rs

```diff
@@ -266,3 +262,45 @@ fn push_schema_definition_changes(
         }),
     }
 }
+
+fn opt_type_eq(a: Option<&ast::Type<'_>>, b: Option<&ast::Type<'_>>) -> bool {
+    match (a, b) {
+        (Some(a), Some(b)) => type_eq(a, b),
+        (None, None) => true,
+        _ => false,
+    }
+}
+
+fn type_eq(a: &ast::Type<'_>, b: &ast::Type<'_>) -> bool {
+    a.name() == b.name() && a.wrappers().eq(b.wrappers())
+}
+
+fn value_eq(a: &ast::Value<'_>, b: &ast::Value<'_>) -> bool {
+    // Same enum variants and equal values
+    match (a, b) {
+        (ast::Value::Int(a), ast::Value::Int(b)) => a == b,
+        (ast::Value::Variable(a), ast::Value::Variable(b)) => a == b,
+        (ast::Value::Float(a), ast::Value::Float(b)) => a == b,
+        (ast::Value::String(a), ast::Value::BlockString(b)) => a == b,
+        (ast::Value::String(a), ast::Value::String(b)) => a == b,
+        (ast::Value::BlockString(a), ast::Value::String(b)) => a == b,
+        (ast::Value::BlockString(a), ast::Value::BlockString(b)) => a == b,
+        (ast::Value::Boolean(a), ast::Value::Boolean(b)) => a == b,
+        (ast::Value::Null, ast::Value::Null) => true,
+        (ast::Value::Enum(a), ast::Value::Enum(b)) => a == b,
+        (ast::Value::List(a), ast::Value::List(b)) => {
+            a.len() == b.len() && a.iter().zip(b.iter()).all(|(a, b)| value_eq(a, b))
+        }
+        (ast::Value::Object(a), ast::Value::Object(b)) => {
+            a.len() == b.len()
+                && a.iter().all(|(name, value)| {
```

Comment:
this is wrong: b could have keys a doesn't have and this would still return true

Reply:
Doesn't the equal length prevent that?

This looks fairly similar to the HashMap impl: https://doc.rust-lang.org/src/std/collections/hash/map.rs.html#1289

Reply:
ah true, should have self-reviewed more carefully